File: utils/crypto.js
import CryptoJS from 'crypto-es';

// Secret keys for HMAC and AES (use the same keys as in your Java code)
const hmacKey = '1234567890123456';
const aesKey = '1234567890123456'; // 16 bytes for AES-128

// Encrypt the message
export const encrypt = (message) => {
    // Generate a random IV (Initialization Vector)
    const iv = CryptoJS.lib.WordArray.random(16); // 16 bytes IV for AES

    // Encrypt the message using AES with the generated IV
    const encrypted = CryptoJS.AES.encrypt(message, CryptoJS.enc.Utf8.parse(aesKey), { iv: iv }).toString();

    // Generate HMAC for the encrypted message
    const hmac = CryptoJS.HmacSHA256(encrypted, hmacKey).toString(CryptoJS.enc.Base64);

    // Combine IV (Base64 encoded), the encrypted message, and the HMAC
    const combined = `${CryptoJS.enc.Base64.stringify(iv)}:${encrypted}:${hmac}`;

    return combined;
};

// Decrypt the message
export const decrypt = (combined) => {
    const [ivBase64, encryptedMessage, receivedHmac] = combined.split(':');

    // Verify HMAC
    const calculatedHmac = CryptoJS.HmacSHA256(encryptedMessage, hmacKey).toString(CryptoJS.enc.Base64);
    if (calculatedHmac !== receivedHmac) {
        throw new Error('Invalid HMAC: The message has been tampered with.');
    }

    // Decrypt the message using AES
    const iv = CryptoJS.enc.Base64.parse(ivBase64); // Parse the IV from Base64
    const decrypted = CryptoJS.AES.decrypt(encryptedMessage, CryptoJS.enc.Utf8.parse(aesKey), { iv: iv });
    const originalMessage = decrypted.toString(CryptoJS.enc.Utf8);
    return originalMessage;
};

File: utils/crypto.java
import javax.crypto.Cipher;
import javax.crypto.KeyGenerator;
import javax.crypto.Mac;
import javax.crypto.SecretKey;
import javax.crypto.spec.SecretKeySpec;
import javax.crypto.spec.IvParameterSpec;
import java.nio.charset.StandardCharsets;
import java.security.SecureRandom;
import java.util.Base64;
 
public class EncryptionUtils {
 
    // Secret keys for HMAC and AES
    private static final String HMAC_KEY = "1234567890123456";
    private static final String AES_KEY = "1234567890123456"; // Must be 16, 24, or 32 bytes
 
    // Encrypt the message
    public static String encrypt(String message) throws Exception {
        // Generate AES SecretKey
        SecretKey aesKey = new SecretKeySpec(AES_KEY.getBytes(StandardCharsets.UTF_8), "AES");
 
        // Initialize Cipher for AES Encryption
        Cipher cipher = Cipher.getInstance("AES/CBC/PKCS5Padding");
        byte[] iv = new byte[cipher.getBlockSize()];
        new SecureRandom().nextBytes(iv);  // Generate IV
        IvParameterSpec ivParams = new IvParameterSpec(iv);
        cipher.init(Cipher.ENCRYPT_MODE, aesKey, ivParams);
 
        // Encrypt the message
        byte[] encryptedBytes = cipher.doFinal(message.getBytes(StandardCharsets.UTF_8));
        String encryptedMessage = Base64.getEncoder().encodeToString(encryptedBytes);
 
        // Generate HMAC for the encrypted message
        Mac mac = Mac.getInstance("HmacSHA256");
        SecretKey hmacKey = new SecretKeySpec(HMAC_KEY.getBytes(StandardCharsets.UTF_8), "HmacSHA256");
        mac.init(hmacKey);
        byte[] hmacBytes = mac.doFinal(encryptedMessage.getBytes(StandardCharsets.UTF_8));
        String hmac = Base64.getEncoder().encodeToString(hmacBytes);
 
        // Combine the encrypted message and the HMAC (including IV as a prefix)
        String combined = Base64.getEncoder().encodeToString(iv) + ":" + encryptedMessage + ":" + hmac;
        return combined;
    }
 
    // Decrypt the message
    public static String decrypt(String combined) throws Exception {
        String[] parts = combined.split(":");
        if (parts.length != 3) {
            throw new IllegalArgumentException("Invalid input format.");
        }
 
        // Extract IV, encrypted message, and HMAC
        byte[] iv = Base64.getDecoder().decode(parts[0]);
        String encryptedMessage = parts[1];
        String receivedHmac = parts[2];
 
        // Verify HMAC
        Mac mac = Mac.getInstance("HmacSHA256");
        SecretKey hmacKey = new SecretKeySpec(HMAC_KEY.getBytes(StandardCharsets.UTF_8), "HmacSHA256");
        mac.init(hmacKey);
        byte[] calculatedHmacBytes = mac.doFinal(encryptedMessage.getBytes(StandardCharsets.UTF_8));
        String calculatedHmac = Base64.getEncoder().encodeToString(calculatedHmacBytes);
 
        if (!calculatedHmac.equals(receivedHmac)) {
            throw new Exception("Invalid HMAC: The message has been tampered with.");
        }
 
        // Decrypt the message using AES
        SecretKey aesKey = new SecretKeySpec(AES_KEY.getBytes(StandardCharsets.UTF_8), "AES");
        Cipher cipher = Cipher.getInstance("AES/CBC/PKCS5Padding");
        IvParameterSpec ivParams = new IvParameterSpec(iv);
        cipher.init(Cipher.DECRYPT_MODE, aesKey, ivParams);
 
        byte[] decryptedBytes = cipher.doFinal(Base64.getDecoder().decode(encryptedMessage));
        return new String(decryptedBytes, StandardCharsets.UTF_8);
    }
 
    // Example usage
    public static void main(String[] args) {
        try {
            String message = "Affan";
            String encryptedMessage = encrypt(message);
            System.out.println("Encrypted Message with HMAC: " + encryptedMessage);
 
            String decryptedMessage = decrypt(encryptedMessage);
            System.out.println("Decrypted Message: " + decryptedMessage);
 
        } catch (Exception e) {
            e.printStackTrace();
            System.err.println("Decryption failed: " + e.getMessage());
        }
    }
}